File: feishu-integration/scripts/msg.py
#!/usr/bin/env python3
"""Feishu Messaging CLI — Send messages, manage chats.

Usage via unified CLI:
  ./feishu msg send|chats|history
"""

import argparse
import json
import os
import re
import sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, SCRIPT_DIR)

from feishu_api import FeishuClient, Log, output


# ─── Helpers ─────────────────────────────────────────────────────────────────

def _resolve_receive_id(client: FeishuClient, user: str = "", chat_id: str = "") -> tuple:
    """Resolve target to (receive_id, receive_id_type).

    Priority: chat_id > user name resolution.
    Returns (id, type) where type is 'chat_id' or 'open_id'.
    """
    if chat_id:
        return chat_id, "chat_id"
    if user:
        from members import resolve_member
        open_id = resolve_member(user, client)
        if not open_id:
            Log.error(f"Cannot resolve user '{user}'")
            sys.exit(1)
        return open_id, "open_id"
    Log.error("Specify --chat-id or --user")
    sys.exit(1)


def _build_text_content(text: str) -> tuple:
    """Build text message content. Returns (msg_type, content_json)."""
    return "text", json.dumps({"text": text})


def _build_interactive_content(card_path: str) -> tuple:
    """Build interactive card content from JSON file. Returns (msg_type, content_json)."""
    with open(card_path) as f:
        card = json.load(f)
    return "interactive", json.dumps(card)


def _build_post_content(title: str, text: str) -> tuple:
    """Build rich-text (post) message with inline formatting.

    Supports **bold** and [link text](url) syntax in text content.
    Each line becomes a paragraph in the Feishu post structure.
    """
    # Split text into lines, each becomes a paragraph in post
    lines = text.split("\n") if text else []
    paragraphs = []
    for line in lines:
        elements = []
        # Parse **bold** and [link](url)
        parts = re.split(r'(\*\*[^*]+\*\*|\[[^\]]+\]\([^)]+\))', line)
        for part in parts:
            if part.startswith("**") and part.endswith("**"):
                elements.append({"tag": "text", "text": part[2:-2], "style": ["bold"]})
            elif part.startswith("[") and "]" in part and "(" in part:
                m = re.match(r'\[([^\]]+)\]\(([^)]+)\)', part)
                if m:
                    elements.append({"tag": "a", "text": m.group(1), "href": m.group(2)})
                else:
                    elements.append({"tag": "text", "text": part})
            elif part:
                elements.append({"tag": "text", "text": part})
        if elements:
            paragraphs.append(elements)

    content = {
        "zh_cn": {
            "title": title,
            "content": paragraphs,
        }
    }
    return "post", json.dumps(content)


# ─── CLI ─────────────────────────────────────────────────────────────────────

def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(prog="msg.py", description="Feishu Messaging")
    sub = parser.add_subparsers(dest="command")

    # Send message
    p = sub.add_parser("send", help="Send a message to chat or user")
    p.add_argument("--chat-id", default="", help="Target chat ID (oc_xxx)")
    p.add_argument("--user", default="", help="Target user name (resolved via member cache)")
    p.add_argument("--text", default="", help="Text message content")
    p.add_argument("--post", default="", help="Rich-text message content (supports **bold** and [link](url))")
    p.add_argument("--title", default="", help="Title for post/rich-text messages")
    p.add_argument("--card", default="", help="Path to interactive card JSON file")
    p.add_argument("--reply-to", default="", help="Message ID to reply to")

    # List chats
    p = sub.add_parser("chats", help="List joined chats/groups")
    p.add_argument("--name", default="", help="Filter by chat name keyword")
    p.add_argument("--json", action="store_true", dest="json_output", help="Output as JSON")

    # Get chat info
    p = sub.add_parser("chat-info", help="Get chat details")
    p.add_argument("--chat-id", required=True, help="Chat ID")

    # Message history
    p = sub.add_parser("history", help="Get message history of a chat")
    p.add_argument("--chat-id", required=True, help="Chat ID")
    p.add_argument("--count", type=int, default=20, help="Number of messages (default: 20)")
    p.add_argument("--json", action="store_true", dest="json_output", help="Output as JSON")

    return parser


def main():
    parser = build_parser()
    args = parser.parse_args()

    if not args.command:
        parser.print_help()
        sys.exit(1)

    client = FeishuClient()

    if args.command == "send":
        # Resolve target
        receive_id, receive_id_type = _resolve_receive_id(
            client, user=args.user, chat_id=args.chat_id
        )

        # Build content
        if args.card:
            msg_type, content = _build_interactive_content(args.card)
        elif args.post:
            msg_type, content = _build_post_content(args.title or "", args.post)
        elif args.text:
            msg_type, content = _build_text_content(args.text)
        else:
            Log.error("Specify --text, --post, or --card")
            sys.exit(1)

        # Send
        body = {
            "receive_id": receive_id,
            "msg_type": msg_type,
            "content": content,
        }
        if args.reply_to:
            body["reply_message_id"] = args.reply_to

        params = {"receive_id_type": receive_id_type}
        result = client.post("/im/v1/messages", body, params=params)

        code = result.get("code", -1)
        if code == 0:
            msg_id = result.get("data", {}).get("message_id", "?")
            Log.ok(f"Message sent (id: {msg_id})")
        else:
            Log.error(f"Send failed: {result.get('msg', '?')} (code: {code})")
            output(result)
            sys.exit(1)

    elif args.command == "chats":
        chats = client.get_all("/im/v1/chats", params={
            "page_size": "50",
        })

        if args.name:
            keyword = args.name.lower()
            chats = [c for c in chats if keyword in c.get("name", "").lower()]

        if args.json_output:
            output({"chats": chats, "count": len(chats)})
        else:
            if not chats:
                Log.warn("No chats found")
                sys.exit(0)
            for i, c in enumerate(chats, 1):
                name = c.get("name", "(unnamed)")
                chat_id = c.get("chat_id", "?")
                member_count = c.get("user_count", "?")
                chat_type = c.get("chat_type", "?")
                print(f"  {i:2}. {name}")
                print(f"       id: {chat_id}  type: {chat_type}  members: {member_count}")
            print(f"\n  Total: {len(chats)}")

    elif args.command == "chat-info":
        result = client.get(f"/im/v1/chats/{args.chat_id}")
        output(result)

    elif args.command == "history":
        params = {
            "page_size": str(min(args.count, 50)),
        }
        result = client.get(f"/im/v1/messages", params={
            **params,
            "container_id_type": "chat",
            "container_id": args.chat_id,
        })
        items = result.get("data", {}).get("items", [])

        if args.json_output:
            output({"messages": items, "count": len(items)})
        else:
            if not items:
                Log.warn("No messages found")
                sys.exit(0)
            for msg in items:
                sender_id = msg.get("sender", {}).get("id", "?")
                msg_type = msg.get("msg_type", "?")
                create_time = msg.get("create_time", "?")
                body_raw = msg.get("body", {}).get("content", "{}")
                try:
                    body = json.loads(body_raw)
                    text = body.get("text", body_raw[:100])
                except (json.JSONDecodeError, AttributeError):
                    text = str(body_raw)[:100]
                print(f"  [{create_time}] ({msg_type}) {sender_id}: {text}")
            print(f"\n  Messages: {len(items)}")

    else:
        parser.print_help()


if __name__ == "__main__":
    main()

File: feishu-integration/scripts/feishu-ts/src/msg.ts
/**
 * Feishu Messaging CLI — Send messages, manage chats.
 *
 * Usage: ./feishu msg send|chats|chat-info|history
 */

import * as fs from "node:fs";
import * as lark from "@larksuiteoapi/node-sdk";
import { Log } from "./log.js";
import { createLarkClient, userRequest, getAllPages } from "./client.js";
import { parseArgs, respond, fail } from "./utils.js";
import { resolveMember } from "./members.js";

// ── Helpers ─────────────────────────────────────────────────────────────────

async function resolveReceiveId(
  client: lark.Client,
  user: string,
  chatId: string,
  chatName: string
): Promise<{ id: string; type: string }> {
  if (chatId) return { id: chatId, type: "chat_id" };

  // Compound: resolve chat by name
  if (chatName) {
    const chats = await getAllPages(client, "/im/v1/chats", {
      page_size: "50",
    }) as Record<string, unknown>[];
    const keyword = chatName.toLowerCase();
    const matched = chats.filter(c =>
      ((c.name ?? "") as string).toLowerCase().includes(keyword)
    );
    if (!matched.length) fail(`No chat matching '${chatName}'`);
    if (matched.length > 1) {
      const names = matched.map(c => ({ name: c.name, chat_id: c.chat_id }));
      fail(
        `Multiple chats match '${chatName}' (${matched.length}). Specify --chat-id.`,
        JSON.stringify(names)
      );
    }
    const id = (matched[0]!.chat_id ?? "") as string;
    Log.ok(`Resolved chat: ${matched[0]!.name} → ${id}`);
    return { id, type: "chat_id" };
  }

  if (user) {
    const openId = await resolveMember(user, client);
    if (!openId) fail(`Cannot resolve user '${user}'`);
    return { id: openId, type: "open_id" };
  }
  fail("Specify --chat-id, --chat-name, or --user");
}

function buildPostContent(
  title: string,
  text: string
): { msgType: string; content: string } {
  const lines = text ? text.split("\\n") : [];
  const paragraphs: Record<string, unknown>[][] = [];

  for (const line of lines) {
    const elements: Record<string, unknown>[] = [];
    const parts = line.split(/(\*\*[^*]+\*\*|\[[^\]]+\]\([^)]+\))/);
    for (const part of parts) {
      if (part.startsWith("**") && part.endsWith("**")) {
        elements.push({
          tag: "text",
          text: part.slice(2, -2),
          style: ["bold"],
        });
      } else if (part.startsWith("[") && part.includes("](")) {
        const m = part.match(/\[([^\]]+)\]\(([^)]+)\)/);
        if (m) {
          elements.push({ tag: "a", text: m[1], href: m[2] });
        } else {
          elements.push({ tag: "text", text: part });
        }
      } else if (part) {
        elements.push({ tag: "text", text: part });
      }
    }
    if (elements.length) paragraphs.push(elements);
  }

  const content = {
    zh_cn: { title, content: paragraphs },
  };
  return { msgType: "post", content: JSON.stringify(content) };
}

// ── CLI Router ──────────────────────────────────────────────────────────────

export async function msgMain(argv: string[]): Promise<void> {
  const { command, args } = parseArgs(argv);
  const client = createLarkClient();

  if (command === "send") {
    const { id: receiveId, type: receiveIdType } = await resolveReceiveId(
      client,
      args.user ?? "",
      args["chat-id"] ?? "",
      args["chat-name"] ?? ""
    );

    let msgType: string;
    let content: string;

    if (args.card) {
      const cardJson = JSON.parse(fs.readFileSync(args.card, "utf-8"));
      msgType = "interactive";
      content = JSON.stringify(cardJson);
    } else if (args.post) {
      const result = buildPostContent(args.title ?? "", args.post);
      msgType = result.msgType;
      content = result.content;
    } else if (args.text) {
      msgType = "text";
      content = JSON.stringify({ text: args.text });
    } else {
      fail("Specify --text, --post, or --card");
    }

    const body: Record<string, unknown> = {
      receive_id: receiveId,
      msg_type: msgType!,
      content: content!,
    };
    if (args["reply-to"]) body.reply_message_id = args["reply-to"];

    const result = await userRequest(client, "POST", "/im/v1/messages", body, {
      receive_id_type: receiveIdType,
    });

    const code = result.code as number;
    if (code === 0) {
      const msgId = ((result.data as Record<string, unknown>)?.message_id ?? "?") as string;
      respond({ message_id: msgId, receive_id: receiveId }, `Message sent (id: ${msgId})`);
    } else {
      fail(`Send failed: ${result.msg ?? "?"} (code: ${code})`);
    }

  } else if (command === "chats") {
    let chats = (await getAllPages(client, "/im/v1/chats", {
      page_size: "50",
    })) as Record<string, unknown>[];

    if (args.name) {
      const keyword = args.name.toLowerCase();
      chats = chats.filter((c) =>
        ((c.name ?? "") as string).toLowerCase().includes(keyword)
      );
    }

    respond({
      chats: chats.map(c => ({
        chat_id: c.chat_id,
        name: c.name,
        chat_type: c.chat_type,
        user_count: c.user_count,
      })),
      count: chats.length,
    }, `${chats.length} chat(s)`);

  } else if (command === "chat-info") {
    if (!args["chat-id"]) fail("--chat-id required");
    const result = await userRequest(client, "GET", `/im/v1/chats/${args["chat-id"]}`);
    if ((result.code as number) !== 0) fail(`Chat info failed: ${result.msg ?? "?"}`);
    respond(result.data);

  } else if (command === "history") {
    if (!args["chat-id"]) fail("--chat-id required");
    const count = Math.min(Number(args.count ?? "20"), 50);
    const result = await userRequest(client, "GET", "/im/v1/messages", undefined, {
      page_size: String(count),
      container_id_type: "chat",
      container_id: args["chat-id"],
    });
    const items = ((result.data as Record<string, unknown>)?.items ??
      []) as Record<string, unknown>[];

    respond({
      messages: items.map(msg => {
        const bodyRaw = ((msg.body as Record<string, unknown>)?.content ?? "{}") as string;
        let text: string;
        try {
          const body = JSON.parse(bodyRaw);
          text = (body.text ?? bodyRaw.slice(0, 200)) as string;
        } catch {
          text = String(bodyRaw).slice(0, 200);
        }
        return {
          message_id: msg.message_id,
          sender_id: (msg.sender as Record<string, unknown>)?.id,
          msg_type: msg.msg_type,
          create_time: msg.create_time,
          text,
        };
      }),
      count: items.length,
    }, `${items.length} message(s)`);

  } else {
    respond({
      commands: ["send", "chats", "chat-info", "history"],
      usage: "./feishu msg <command>",
    }, "Messaging");
  }
}
